src/amount.rs

```diff
@@ -353,48 +437,172 @@ impl Amount {
+        }
+    }
+
+    // attempts to parse a string in the notation: x*10^y
+    // where x must be positive and y may be negative.
+    fn from_str_notation_full(s: &str) -> Result<Self> {
+        let parts: Vec<&str> = s.split(|c| c == '*' || c == '^').collect();
+
+        if parts.len() != 3 {
+            return Err(Error::AmountUnparseable);
+        }
+        if parts[1] != "10" {
+            return Err(Error::AmountUnparseable);
+        }
+
+        let count: AmountCounter = parts[0].parse().map_err(|_| Error::AmountUnparseable)?;
+        let unit: PowerOfTen = parts[2].parse().map_err(|_| Error::AmountUnparseable)?;
+
+        // Self::new will verify that count <= Self::counter_max()
+        Self::new(count, unit)
+    }
+
+    // attempts to parse a string in the notation: 10^y
+    // where y may be negative.
+    fn from_str_notation_short(s: &str) -> Result<Self> {
+        let parts: Vec<&str> = s.split('^').collect();
+
+        if parts[0] != "10" {
+            return Err(Error::AmountUnparseable);
+        }
+        if parts.len() != 2 {
+            return Err(Error::AmountUnparseable);
+        }
+        let unit: PowerOfTen = parts[1].parse().map_err(|_| Error::AmountUnparseable)?;
+
+        // Self::new will verify that count <= Self::counter_max()
+        Self::new(1, unit)
+    }
+
+    // attempts to parse an SI string like "253 yotta"
+    fn from_str_si(s: &str) -> Result<Self> {
+        let map = Self::si_map();
+
+        let parts: Vec<&str> = s.split(' ').collect();
+        if parts.len() != 2 {
+            return Err(Error::AmountUnparseable);
+        }
+
+        let count: AmountCounter = parts[0].parse().map_err(|_| Error::AmountUnparseable)?;
+        let name = parts[1].to_lowercase();
+
+        let (unit, _) = map
+            .iter()
+            .find(|(_, v)| **v == name)
+            .ok_or(Error::AmountUnparseable)?;
+
+        Self::new(count, *unit)
+    }
+}
+
+impl FromStr for Amount {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // parses eg 2345322200000000
+        if let Ok(v) = Self::from_str_u128(s) {
+            return Ok(v);
+        }
+
+        // parses eg 3*10^-5
+        if let Ok(v) = Self::from_str_notation_full(s) {
+            return Ok(v);
+        }
+
+        // parses eg 10^-5
+        if let Ok(v) = Self::from_str_notation_short(s) {
+            return Ok(v);
+        }
+
+        // parses eg 253 yotta
+        if let Ok(v) = Self::from_str_si(s) {
+            return Ok(v);
+        }
+
+        Err(Error::AmountUnparseable)
+    }
+}
+
+impl TryFrom<u128> for Amount {
+    type Error = Error;
+
+    fn try_from(n: u128) -> Result<Self, Self::Error> {
+        let (unit, count) = calc_exponent_u128(n).ok_or(Error::AmountUnparseable)?;
+
+        let count = AmountCounter::try_from(count).map_err(|_| Error::AmountUnparseable)?;
+        Amount::new(count, unit)
+    }
+}
```

Comment:
This is just a thought, haven't tested it, but you may be able to get all of this `TryFrom` impls for free with:

```rust
impl<T: Into<u128>> TryFrom<T> for Amount {
    type Error = Error;

    fn try_from(n: T) -> Result<Self, Self::Error> {
        let (unit, count) = calc_exponent_u128(n.into::<u128>()).ok_or(Error::AmountUnparseable)?;

        let count = AmountCounter::try_from(count).map_err(|_| Error::AmountUnparseable)?;
        Amount::new(count, unit)
    }
}
```

Reply:
ok, I guess we'll leave it as-is then.